meetings/working-groups/collection-literals/CL-2023-08-03.md

```diff
@@ -0,0 +1,140 @@
+
+
+Today, a user can write code like so:
+
+```c#
+ReadOnlySpan<int> x1 = stackalloc int[] { a, b, c }; // stack allocated, local scope.
+ReadOnlySpan<int> x2 = new int[] { a, b, c };        // heap allocated, global scope.
+```
+
+Based on how the variable is initialized, the language then dictates rules about how it can be used.  For example, with the above code it would be illegal to `return x1;` from the method (since it has local scope), while it would be fine to `return x2;`.  Variables without an initializer are assumed to have global scope.  So the following, for example, would be illegal:
+
+```c#
+ReadOnlySpan<int> x3;              // global scope
+x3 = stackalloc int[] { a, b, c }; // illegal.  global scope now points at local data.
+```
+
+Unlike existing constructs, which clearly delineate whether they are on the stack or heap, a decision needs to be made as to what the following means:
+
+```c#
+ReadOnlySpan<int> x4 = [a, b, c];
+```
+
+The working group considered several options.
+
+### Option 1.  Stack-alloc'ed only if target has local-scope.
+
+The first option considered was that a collection expression used with a `ReadOnlySpan<T>` would itself only be stack-alloc'ed if the span itself had an explicitly stated local scope.   In other words, the following:
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c];        // heap allocated, global scope.
+scoped ReadonlySpan<int> x2 = [a, b, c]; // stack allocated, local scope.
+```
+
+This has the positive that both heap and local allocations are possible, and choosing between them is as simple as adding `scoped` or not.  Unfortunately it has several major negatives the working group was quite dissatisfied with:
+
+1. We believe the *common* case for using spans with collection expressions will be for people wanting stack allocated data.  Having the default (non-`scoped`) syntax basically flips that and makes the common case more cluttered and less pleasant to work with.
+2. We believe the standard intuition for nearly all users (including experts in the low-level) space would be that the first line produces stack-allocated data.  After all, that's the most common reason to create locals of span types, and their belief would be that "of course if i initialized with a collection expression" that "it will do the right thing".  
+3. We believe that the first line would actually be a *large* "pit of fail" for users.  It would be very easy and attractive for customers to just take existing code, hear about collection expressions (including that it works for spans) and translate `stackalloc` code over to just using the collection expression.  This would then result in near immediate performance tanking due to potentially extreme allocations happening in hotspots that didn't exist before.  Such problems could even easily lead to this feature being entirely banned in codebases due to the fear of this happening.
```

Comment:
Perhaps:
```suggestion
3. We believe that the first line would actually be a *large* "pit of fail" for users.  It would be very easy and attractive for customers to just take existing code, hear about collection expressions (including that it works for spans) and translate `stackalloc` code over to just using the collection expression, resulting in allocations that didn't exist before.
```